connectors/sources/network_drive.py

```diff
@@ -414,54 +416,120 @@ async def traverse_diretory(self, path):
                 self._logger.exception(
                     f"Error while scanning the path {current_path}. Error {exception}"
                 )
-            for file in directory_info:
-                yield self.format_document(file=file)
-                if file.is_dir():
-                    stack.append(file.path)
-
-    async def get_directory_details(self):
-        self._logger.debug("Fetching the directory tree from remote server")
-        paths = await asyncio.to_thread(
-            partial(
-                smbclient.walk,
-                top=rf"\\{self.server_ip}/{self.drive_path}",
-                port=self.port,
-            ),
+                continue
+
+    def is_match_with_previous_rules(
+        self, file_path, indexed_rules, match_with_previous_rules
+    ):
+        # Check if the file is matched with any of the previous indexed rules
+        for indexed_rule in indexed_rules:
+            if not match_with_previous_rules:
+                match_with_previous_rules = glob.globmatch(
+                    file_path, indexed_rule, flags=glob.GLOBSTAR
+                )
+                if match_with_previous_rules:
+                    break
+        return match_with_previous_rules
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=[SMBOSError, SMBException],
+    )
+    async def traverse_directory_for_syncrule(self, path, glob_pattern, indexed_rules):
+        self._logger.debug(
+            "Fetching the directory tree from remote server and content of directory on path"
         )
-        return list(paths)
+        stack = deque()
+        stack.append(path)
+
+        while stack:
+            directory_info = []
+            current_path = stack.pop()
+            try:
+                directory_info = await asyncio.to_thread(
+                    partial(
+                        smbclient.scandir,
+                        path=current_path,
+                        port=self.port,
+                    ),
+                )
+                for file in directory_info:
+                    match_with_previous_rules = False
+                    if file.is_dir():
+                        stack.append(file.path)
+                    file_path = file.path.split("/", 1)[1].replace("\\", "/")
+                    is_file_match = glob.globmatch(
+                        file_path, glob_pattern, flags=glob.GLOBSTAR
+                    )
+                    match_with_previous_rules = self.is_match_with_previous_rules(
+                        file_path, indexed_rules, match_with_previous_rules
+                    )
```

Comment:
This function is to not yield documents twice if multiple rules trigger for one object. It will fetch all the file folders for first pattern and from next pattern we'll match files from the previous rules and if it match with any of the previous rules then we'll not index it as it was already indexed while running previous patterns.